homeassistant/components/esphome/config_flow.py

```diff
@@ -149,93 +149,43 @@ async def async_step_zeroconf(
         self, discovery_info: zeroconf.ZeroconfServiceInfo
     ) -> FlowResult:
         """Handle zeroconf discovery."""
-        # Hostname is format: livingroom.local.
-        local_name = discovery_info.hostname[:-1]
-        node_name = local_name[: -len(".local")]
-        address = discovery_info.properties.get("address", local_name)
+        mac_address: str | None = discovery_info.properties.get("mac")
 
-        # Check if already configured
-        await self.async_set_unique_id(node_name)
-        self._abort_if_unique_id_configured(updates={CONF_HOST: discovery_info.host})
-
-        for entry in self._async_current_entries():
-            already_configured = False
-
-            if CONF_HOST in entry.data and entry.data[CONF_HOST] in (
-                address,
-                discovery_info.host,
-            ):
-                # Is this address or IP address already configured?
-                already_configured = True
-            elif DomainData.get(self.hass).is_entry_loaded(entry):
-                # Does a config entry with this name already exist?
-                data = DomainData.get(self.hass).get_entry_data(entry)
-
-                # Node names are unique in the network
-                if data.device_info is not None:
-                    already_configured = data.device_info.name == node_name
-
-            if already_configured:
-                # Backwards compat, we update old entries
-                if not entry.unique_id:
-                    self.hass.config_entries.async_update_entry(
-                        entry,
-                        data={
-                            **entry.data,
-                            CONF_HOST: discovery_info.host,
-                        },
-                        unique_id=node_name,
-                    )
-
-                return self.async_abort(reason="already_configured")
+        # Mac address was added in Sept 20, 2021.
+        # https://github.com/esphome/esphome/pull/2303
+        if mac_address is None:
+            return self.async_abort(reason="mdns_missing_mac")
+
+        # mac address is lowercase and without :, normalize it
+        mac_address = ":".join(
+            mac_address[i : i + 2].upper() for i in range(0, len(mac_address), 2)
+        )
```

Comment:
I think you could use `format_mac` from `device_registry`

Reply:
ha I was looking for that one but couldn't find it with grepping. Had CoPilot write this one 😅

Reply:
yeah the codebase seems to have expanded a bit 😉